core/include/definitions/algebra.hpp

```diff
@@ -17,41 +19,59 @@
 namespace traccc
 {
 
+#ifdef __CUDACC__      
+  __device__ __host__
+#endif      
```

Comment:
You'll probably want to factor out those repeating three lines into a macro.

Reply:
Wait... does this actually work? 😕  I was convinced that not even `std::array` could be used directly in device code. Can it? 😕  (I don't know, maybe the latest CUDA version allows for this?...)

Similar to [llama::Array](https://github.com/alpaka-group/llama/blob/develop/include/llama/Array.hpp), I added [vecmem::static_vector](https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/containers/static_vector.hpp) exactly for this type of usage. To handle small objects that are created directly in device code. (And can even be passed from the host to the device or vice-versa as they are.) I've had the plan for a while to introduce a `vecmem::static_array` as well, which would pretty much just be a simplified version of `vecmem::static_vector`...

Reply:
Hmm... Apparently CUDA 11.2 (didn't try older versions...) **can** be forced into using [std::array](https://en.cppreference.com/w/cpp/container/array). :thinking: I wrote this little test for it:

```cuda
#include <cuda_runtime.h>

#include <array>
#include <iostream>
#include <vector>

__global__
void arrayTestKernel( int size, int* ptr ) {

   const int i = blockIdx.x * blockDim.x + threadIdx.x;
   if( i >= size ) {
      return;
   }

   const std::array< int, 2 > constants = { 2, 3 };
   ptr[ i ] = ptr[ i ] * constants.at( 0 ) + constants.back();
   return;
}

int main() {

   std::vector< int > host_vec = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
   int* device_array = nullptr;
   cudaMalloc( &device_array, host_vec.size() * sizeof( int ) );
   cudaMemcpy( device_array, host_vec.data(), host_vec.size() * sizeof( int ),
               cudaMemcpyHostToDevice );

   arrayTestKernel<<< 1, host_vec.size() >>>( host_vec.size(), device_array );

   cudaMemcpy( host_vec.data(), device_array, host_vec.size() * sizeof( int ),
               cudaMemcpyDeviceToHost );
   cudaFree( device_array );

   const std::vector< int > expected_result = {
      5, 7, 9, 11, 13, 15, 17, 19, 21, 23 };
   if( host_vec != expected_result ) {
      std::cerr << "What?!?" << std::endl;
      return 1;
   }
   std::cout << "All good!" << std::endl;

   return 0;
}
```

Which works like:

```
[bash][Elrond]:test > nvcc -o test --expt-relaxed-constexpr array_test.cu 
[bash][Elrond]:test > ./test 
All good!
[bash][Elrond]:test >
```

So even with CUDA 11.2 one needs to use an experimental feature for this. But it does work... So it's a nay on the `vecmem::static_array` implementation... :stuck_out_tongue:

P.S. [std::vector](https://en.cppreference.com/w/cpp/container/vector) is of course still a no-go in device code.

```
/data/projects/vecmem/vecmem/tests/cuda/array_test.cu(16): error: calling a __host__ function("std::allocator<int> ::allocator") from a __global__ function("arrayTestKernel") is not allowed

/data/projects/vecmem/vecmem/tests/cuda/array_test.cu(16): error: identifier "std::allocator<int> ::allocator" is undefined in device code

/data/projects/vecmem/vecmem/tests/cuda/array_test.cu(16): error: calling a __host__ function("std::vector<int, ::std::allocator<int> > ::vector") from a __global__ function("arrayTestKernel") is not allowed

/data/projects/vecmem/vecmem/tests/cuda/array_test.cu(16): error: identifier "std::vector<int, ::std::allocator<int> > ::vector" is undefined in device code

/data/projects/vecmem/vecmem/tests/cuda/array_test.cu(16): error: calling a __host__ function("std::allocator<int> ::~allocator") from a __global__ function("arrayTestKernel") is not allowed
...
```

So `vecmem::static_vector` will continue having a use-case. :wink:

Reply:
@HadrienG2 Sure. I will do it
@krasznaa Good point. I used the `--expt-relaxed-constexpr` to make it work. 
The general discussion point is whether to write `algebra.hpp` for cpu and cuda separately. If we want to separate them, the featurer of `static::array` or `static::vector` (or even `static::tuple`) will be definitely useful.